### ISSUE TITLE
add dedicated support for pydantic

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#2403]

**Comment by rayzeller** [comment]
## Progress on #2719 (Validator Classmethods)

I originally raised #2719 (validator decorators as implicit classmethods) and @sharkdp raised some design questions there about how ty should approach third-party library detection. Wanted to share my research here on the parent issue in case it's useful for the broader Pydantic support effort.

### The problem

Pydantic v2 [documents](https://docs.pydantic.dev/latest/concepts/validators/) that `@field_validator` validators are "class methods" and recommends adding `@classmethod` for type checkers. ty treats these as instance methods:

```python
@field_validator("name")
def validate_name(cls, v: str) -> str:  # ty infers (self: User, v: str) — wrong
    return v
```

### Proposed approach

**Scope:** Start narrow — `field_validator` + `field_serializer` only. `model_validator` has a `mode="after"` edge case ([mypy.py#L504-L511](https://github.com/pydantic/pydantic/blob/main/pydantic/mypy.py#L504-L511)) that needs separate handling.

**Detection:** Module-qualified symbol detection — check that the resolved callable is named `field_validator`/`field_serializer` AND comes from `pydantic.*`. Handles import aliasing since we inspect the resolved `FunctionType`.

Implementation note: `KnownModule` is stdlib-only ([module.rs#L375-L381](https://github.com/astral-sh/ruff/blob/main/crates/ty_module_resolver/src/module.rs#L375-L381)), so this uses `file_to_module()` instead. Is there a preferred pattern for third-party detection?

### Design questions (from #2719)

1. **Config or always-on?** Proposing always-on — we're implementing Pydantic's documented contract. Happy to add a toggle if preferred.

2. **Version scope?** Pydantic v2 only (v1 is maintenance mode with different internals).

I have a working prototype. Happy to open a draft PR or adapt based on feedback.

**Comment by carljm** [comment]
My opinion is that we should expand `KnownModule` to be usable for third-party modules like pydantic. I think that we should still keep it as strict as possible; that is, we should specially mark known-modules that we expect to find in third-party rather than in stdlib, and we should require that we find each known-module in the type of path that we expect to find it in (i.e. stdlib for most, third-party for some, never first-party).

Beyond that I think detection by module name is fine. If someone creates a third-party module named "pydantic" which is not actually pydantic, and yet has the same known classes/functions that we expect to find in pydantic -- well then, they get our special handling for pydantic. I think we can live with that. (If it's ever a problem in the real world, we can implement additional detection based on some known top-level constants in pydantic or something.)

I think the special handling for pydantic should be on-by-default -- I don't see a use case for disabling it.

**Comment by Viicos** [comment]
Great to see `ty` gaining Pydantic support! If it can help, I initially provided feedback for Pyrefly [here](https://github.com/facebook/pyrefly/discussions/854#discussioncomment-14063403), it might be useful for you too. By special casing Pydantic in `ty`, you might encounter edge cases worth looking into.

In some places, Pydantic has questionable behavior that will or might change in the future (e.g. the next V3 release). I'll provide some examples here, and update the list as I find more examples. For behavior that is expected to changes between versions, if you still want to support them, I believe it would be possible for `ty` to read the `pydantic.VERSION` module attribute (it is guaranteed to be a [spec compliant](https://packaging.python.org/en/latest/specifications/version-specifiers/) specifier). I would also suggest _not_ supporting any V1 specific feature (class-based `Config`, etc).

---

- When using inheritance, configuration of a model is merged with the parent classes. However, things get complicated when multiple bases are present. Here is pseudo-code used to merge configuration together. For any class `cls`, with `cls_kw` being the class arguments [^1]:

    ```python
    final_config = {}
    for base in cls.__bases__:
        final_config.update(getattr(base, 'model_config', {}))

    final_config.update(getattr(cls, 'model_config', {}))
    final_config.update(cls_kw)
    ```

    I don't believe that the class arguments overriding the model config is documented, but will probably not change. However, the merging of configuration from base classes is currently inconsistent as it doesn't follow the MRO and will likely change in V3 (tracked by https://github.com/pydantic/pydantic/issues/9992).

- Pydantic supports having the `Field()` function as `Annotated` metadata. We've got many reports of users getting confused about e.g. `Annotated[int, Field(alias='my_alias')]` not being understood by type checkers. However, it gets more complicated when you have multiple metadata elements (and possibly an assignment too). Examples:

  - `Annotated[int, Field(alias='b', strict=False), Strict(), Field(alias='c')] = Field(alias='d')`
  - `Annotated[int, Field(description='ann')] = Field(description='assign')`
  - `Annotated[int, Ge(0)] = Field(ge=1)`

  Order is sometimes inconsistent, and planned on changing (see https://github.com/pydantic/pydantic/issues/10507).

- Strictness and type coercion (e.g. `'1'` for an `int` is accepted). Type coercion rules can change (see [pydantic/pydantic#12319](https://github.com/pydantic/pydantic/issues/12319), [pydantic/pydantic#9991](https://github.com/pydantic/pydantic/issues/9991)). For some types, supporting wider accepted types might not be trivial. [This is the up to date documentation](https://pydantic.dev/docs/validation/latest/api/pydantic/standard_library_types/) about validation rules. We also have a [conversion table](https://pydantic.dev/docs/validation/latest/concepts/conversion_table) but I'm not sure it is fully up to date/accurate.

- Fields marked as `Final` with a default value are considered as class variables (see https://github.com/pydantic/pydantic/issues/11119).

[^1]: e.g. as in `class Model(BaseModel, extra=..., strict=...): ...`.

**Comment by sharkdp** [comment]
> Great to see `ty` gaining Pydantic support! If it can help, I initially provided feedback for Pyrefly [here](https://github.com/facebook/pyrefly/discussions/854#discussioncomment-14063403), it might be useful for you too.

Thank you! I updated the description in the top-post here to track the current progress. You can always find our most up-to-date specification of Pydantic support in this [Markdown test suite](https://github.com/astral-sh/ruff/blob/main/crates/ty_python_semantic/resources/mdtest/external/pydantic.md) (all of the `# revealed` and `# error` assertions in that file are checked. Lines without an `# error` assertion are guaranteed not to yield diagnostics).

> In some places, Pydantic has questionable behavior that will or might change in the future (e.g. the next V3 release). I'll provide some examples here, and update the list as I find more examples. For behavior that is expected to changes between versions, if you still want to support them, I believe it would be possible for `ty` to read the `pydantic.VERSION` module attribute (it is guaranteed to be a [spec compliant](https://packaging.python.org/en/latest/specifications/version-specifiers/) specifier). I would also suggest _not_ supporting any V1 specific feature (class-based `Config`, etc).

Right, I was not planning on adding V1 support. Reading `pydantic.VERSION` seems easy to do once V3 is out 👍 


> When using inheritance, configuration of a model is merged with the parent classes. However, things get complicated when multiple bases are present.

I think we already model this correctly. The tests for this are [here](https://github.com/astral-sh/ruff/blob/main/crates/ty_python_semantic/resources/mdtest/external/pydantic.md#model-configuration).

> Pydantic supports having the `Field()` function as `Annotated` metadata. We've got many reports of users getting confused about e.g. `Annotated[int, Field(alias='my_alias')]` not being understood by type checkers. However, it gets more complicated when you have multiple metadata elements (and possibly an assignment too) [...] Order is sometimes inconsistent, and planned on changing

Yes, I saw the linked issue. Given the ambiguity when multiple `Field(..)` calls are involved, my plan would be to only add support for the format below, i.e. a single `Field(...)` call in `Annotated[..]`, and no right hand side (see [tests with TODOs here](https://github.com/astral-sh/ruff/blob/acb62cd5bf6085db19e55cdec2da079e05253ee5/crates/ty_python_semantic/resources/mdtest/external/pydantic.md?plain=1#L754-L778)). Do you think that would be too limiting?

```py
field: Annotated[T, Field(...)]
```



> Strictness and type coercion (e.g. `'1'` for an `int` is accepted). Type coercion rules can change (see [pydantic/pydantic#12319](https://github.com/pydantic/pydantic/issues/12319), [pydantic/pydantic#9991](https://github.com/pydantic/pydantic/issues/9991)). 

I think we could support that in ty by providing pydantic-version-specific variants of the [`Lax*` type aliases](https://github.com/astral-sh/ruff/blob/acb62cd5bf6085db19e55cdec2da079e05253ee5/crates/ty_vendored/ty_extensions/pydantic.pyi#L18-L42) with code like [this](https://docs.astral.sh/ty/features/type-system/#reachability-based-on-types) (funny that this uses pydantic as an example 😄). But I think we're good for now since we only support V2, and V3 is not out.

> For some types, supporting wider accepted types might not be trivial.

I documented some limitations of our approach [here](https://github.com/astral-sh/ruff/blob/acb62cd5bf6085db19e55cdec2da079e05253ee5/crates/ty_python_semantic/resources/mdtest/external/pydantic.md?plain=1#L402-L409) and [here](https://github.com/astral-sh/ruff/blob/acb62cd5bf6085db19e55cdec2da079e05253ee5/crates/ty_python_semantic/resources/mdtest/external/pydantic.md?plain=1#L473-L474). Is that what you mean by "supporting wider accepted types", or did you have something else in mind?

> [This is the up to date documentation](https://pydantic.dev/docs/validation/latest/api/pydantic/standard_library_types/) about validation rules. We also have a [conversion table](https://pydantic.dev/docs/validation/latest/concepts/conversion_table) but I'm not sure it is fully up to date/accurate.

I based my tests [in this section](https://github.com/astral-sh/ruff/blob/main/crates/ty_python_semantic/resources/mdtest/external/pydantic.md#lax-mode-is-the-default) on the conversion table, but I can take a look at the first page you mentioned. So far, the only deviation that I found was that `bytes` was actually accepted by Pydantic at runtime in some places where the documentation says it wasn't (I think e.g. `bytes -> ByteSize` or `bytes -> UUID`). I decided that it's probably not important enough to deviate from "the spec" (the conversion table). If the conversion behavior is not up-to-date anymore, it would be great if that table could be updated. It seems like a valuable resource.

> * Fields marked as `Final` with a default value are considered as class variables (see [Fields marked as `Final` with a default value shouldn't be considered as class variables pydantic/pydantic#11119](https://github.com/pydantic/pydantic/issues/11119)).

I wasn't aware of that, I will add it to the list in the first post here.

**Comment by ChenMoFeiJin** [comment]
I initially noticed this through a CI/local mismatch: our CI uses a pinned `ty` version, while my local environment was using the latest release. The code passed locally but failed in CI, which turned out to be due to a `ty` version difference.

The original case was #3947, where some Pydantic secret types are widened to `Any` in lax constructors. While reducing that case, I looked more broadly at synthesized Pydantic constructor input types and found a few focused mismatches. Some are `ty 0.0.57` regressions, while #3945 appears to be a pre-existing modeling issue:

* #3945
* #3946
* #3947
* #3948

Leaving the links here since they seem related to the `BaseModel` constructor-signature and lax/strict-mode work tracked in this umbrella issue.

**Comment by Viicos** [comment]
> Yes, I saw the linked issue. Given the ambiguity when multiple `Field(..)` calls are involved, my plan would be to only add support for the format below, i.e. a single `Field(...)` call in `Annotated[..]`, and no right hand side (see [tests with TODOs here](https://github.com/astral-sh/ruff/blob/acb62cd5bf6085db19e55cdec2da079e05253ee5/crates/ty_python_semantic/resources/mdtest/external/pydantic.md?plain=1#L754-L778)). Do you think that would be too limiting?

One option I see is to support both having multiple `Field()` (or other specific metadata such as `Strict()`) _and_ having an assignment, but consider duplicate metadata as undefined behavior:

- `Annotated[int, Field(alias='test'), Strict()] = Field(default_factory=...)`: fine, unambiguous
- `Annotated[int, Field(alias='b', strict=False), Strict(), Field(alias='c')] = Field(alias='d')`: bad, unambiguous.

After all, there's no reason for users to do so[^1], and `ty` could even error on such duplicate metadata?

---

> I documented some limitations of our approach [here](https://github.com/astral-sh/ruff/blob/acb62cd5bf6085db19e55cdec2da079e05253ee5/crates/ty_python_semantic/resources/mdtest/external/pydantic.md?plain=1#L402-L409) and [here](https://github.com/astral-sh/ruff/blob/acb62cd5bf6085db19e55cdec2da079e05253ee5/crates/ty_python_semantic/resources/mdtest/external/pydantic.md?plain=1#L473-L474). Is that what you mean by "supporting wider accepted types", or did you have something else in mind?

Yes exactly! And sometimes `Any` is probably better than an over-complicated lax type that can cause subtle type checking issues.

---

> I based my tests [in this section](https://github.com/astral-sh/ruff/blob/main/crates/ty_python_semantic/resources/mdtest/external/pydantic.md#lax-mode-is-the-default) on the conversion table, but I can take a look at the first page you mentioned. So far, the only deviation that I found was that `bytes` was actually accepted by Pydantic at runtime in some places where the documentation says it wasn't (I think e.g. `bytes -> ByteSize` or `bytes -> UUID`). I decided that it's probably not important enough to deviate from "the spec" (the conversion table). If the conversion behavior is not up-to-date anymore, it would be great if that table could be updated. It seems like a valuable resource.

Indeed, I'll try to get the table updated soon.

---

Otherwise, while things could change in the next V3 major release, I don't expect ty's Pydantic support to be completely broken. When we'll start providing V3 pre-releases, I'll make sure to communicate with both the ty and Pyrefly team about upcoming changes so that we can ensure a smooth transition.

[^1]: Only case where I can see this happen is if an alias is being reused, e.g. `StrictInt = Annotated[int, Strict()]` reused in a field as `field: Annotated[StrictInt, Field(...)]`. But anyway, this can't reasonably supported by static analysis to begin with.

**Comment by sharkdp** [comment]
> One option I see is to support both having multiple `Field()` (or other specific metadata such as `Strict()`) _and_ having an assignment, but consider duplicate metadata as undefined behavior:

That seems reasonable, yes -- thank you. I will note that down as a future extension in the top post.

(Edit: this is now implemented)

**Comment by sharkdp** [comment]
We merged a series of PRs that implement basic Pydantic support in ty (most of this is available in 0.0.58, the rest will soon be released in 0.0.59):

- We now distinguish between lax and strict mode for Pydantic models and individual fields, for example:
    ```py
	class Job(BaseModel):
	    results: list[dict[str, int]]
	    attempts: int = Field(strict=True)

	# ✅ ty accepts this because Pydantic can coerce the nested "100" → 100	
	Job(results=[{"score": "100"}], attempts=3)

	# ❌ ty rejects this because `None` cannot be coerced to int
	Job(results=[{"score": None}], attempts=3)

	# ❌ ty rejects this because `attempts` validation is strict
	Job(results=[{"score": "100"}], attempts="3")
	```
- ty can parse Pydantic model configurations from class keywords, `ConfigDict(..)` calls, `dict(..)` calls and dictionary literals (`model_dict = {"extra": "forbid"}`). It recognizes options like `strict`, `frozen`, `extra`, `validate_by_name` and `validate_by_alias`. For example:
    ```py
	class Person(BaseModel):
	    model_config = ConfigDict(extra="forbid")
	
	    name: str
	
	Person(name="Alice", age=20)	 # ❌ ty rejects this when `extra="forbid"`
	```
- Model configuration is also inherited and merged with the configs of base classes according to Pydantic's bespoke rules.
- ty can parse field-specific metadata from the `Field(..)` specifier function, even if it's included in `Annotated[..]` metadata:
    ```py
	class User(BaseModel):
	    name: str = Field(alias="full_name")
	    id: Annotated[int, Field(default_factory=get_next_id), Strict()]
	
	# ✅ ty recognizes the `full_name` alias and the fact that `id` has a default
	User(full_name="John Doe")

    # ❌ ty rejects this because `id` validation is strict
    User(full_name="John Doe", id="1")
	```
- In lax mode, the constructor of a model accepts a broader input type. We use custom type aliases like `LaxInt` for better LSP support:
    <img alt="Image" src="https://github.com/user-attachments/assets/800d54f5-bc03-49ed-827c-327703a84036" />
- We model some behavioral differences between dataclasses and Pydantic models, such as the fact that Pydantic models can have fields without defaults after fields with defaults.
- Underscore-prefixed attributes are recognized as private fields
- `RootModel`s are supported
- We special-case `BaseSettings` models and allow their constructors to omit required fields, since they could be supplied by environment variables.

All known user-reported issues have been closed, and all features that I wanted to get into the stable release have been implemented, so I am closing this ticket for now. I opened https://github.com/astral-sh/ty/issues/3959 to track future improvements.

We are interested in your feedback, so please comment here or open new issues if you work with ty on a Pydantic codebase!

**Comment by tiangolo** [comment]
OMG, this is amazing. 🤩
